### ISSUE TITLE
Train a fastspeech2 without teacher model

### ISSUE THREAD

**Issue by Mingrg** [espnet/espnet#2712]
Hi,
Nowadays, I've been trying to train a fastspeech2 without a teacher model(Tacotron2).Just use the MFA.But I have some problem.

In this [issue](https://github.com/espnet/espnet/issues/2632#issuecomment-729746644) you told me that if I prepare the `duratioins`, it can be.
**So all I need to do is to prepare the durations?Is there any other step should I do?**

In this [guide](https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE/tts1#fastspeech2-training).I saw some stage **needs the teacher model dir** like stage 5 and stage 7. For example:
the **tts_exp** in stage 7
```
$ ./run.sh --stage 7 \
    --tts_exp exp/tts_train_raw_phn_tacotron_g2p_en_no_space \
    --inference_args "--use_teacher_forcing true" \
    --test_sets "tr_no_dev dev eval1"
```
the **teacher_dumpdir** and the  **tts_stats_dir** in stage 5
```
$ ./run.sh --stage 5 \
    --train_config conf/tuning/train_fastspeech2.yaml \
    --teacher_dumpdir exp/tts_train_raw_phn_tacotron_g2p_en_no_space/decode_use_teacher_forcingtrue_train.loss.ave \
    --tts_stats_dir exp/tts_train_raw_phn_tacotron_g2p_en_no_space/decode_use_teacher_forcingtrue_train.loss.ave/stats \
    --write_collected_feats true
```
I know the stage 7 is to prepare the durations, but what should I do about the stage 5? I don't have a teacher model dir(a Tacotron2 model dir).How to execute the stage 5 to calculate additional feature (F0 and energy)?

**Comment by kan-bayashi** [comment]
Let us describe the procedure:
1. Create `durations` files for `tr_no_dev`, `dev` and `eval1` set.
2. Put them somewhere in `/path/to/your_duration_dir/{tr_no_dev,dev,eval1}/durations`.
3. Run from `--stage 5` with the following options:
```sh
$ ./run.sh --stage 5 \
    --train_config conf/tuning/train_fastspeech2.yaml \
    --teacher_dumpdir /path/to/your_duration_dir \
    --tts_stats_dir /path/to/your_duration_dir/stats \
    --write_collected_feats true
```
I've never tried, but it should be working.

In `stage 5` with fastspeech2 config, we will calculate token-averaged F0 and energy with pre-calculated durations and their statistics (mean and var). So stage 5 is needed to run.

**Comment by Mingrg** [comment]
> Let us describe the procedure:
> 
> 1. Create `durations` files for `tr_no_dev`, `dev` and `eval1` set.
> 2. Put them somewhere, e.g., `/path/to/your_duration_dir/{tr_no_dev,dev,eval1}/durations`.
> 3. Run from `--stage 5` with the following options:
> 
> ```shell
> $ ./run.sh --stage 5 \
>     --train_config conf/tuning/train_fastspeech2.yaml \
>     --teacher_dumpdir /path/to/your_duration_dir \
>     --tts_stats_dir /path/to/your_duration_dir/stats \
>     --write_collected_feats true
> ```
> 
> I've never tried, but it should be working.
> 
> In `stage 5` with fastspeech2 config, we will calculate token-averaged F0 and energy with pre-calculated durations and their statistics (mean and var). So stage 5 is needed to run.

Thank you so much! I read the code yesterday, and did it the same way, it worked! I just wanted to close the issue, and then you replied to me, HAHA :)
Anyway, Thank you for answering my questions patiently!

**Comment by Mingrg** [comment]
By the way, I have another question:
In training fastspeech2 model, which one is better to create durations for final speech quality? **use teacher model** or **use MFA**?
Maybe you tried that before, I don't know which one is better, so I want to try.

**Comment by kan-bayashi** [comment]
Actually, I've never compared.
But maybe HMM-based alignment is more precise, especially for the clean speech.
So the MFA will give better quality but the improvement is not so large, I think.
If you share your experiment, that is great :)

**Comment by Mingrg** [comment]
> Actually, I've never compared.
> But maybe HMM-based alignment is more precise, especially for the clean speech.
> So the MFA will give better quality but the improvement is not so large, I think.
> If you share your experiment, that is great :)

Thanks, I will

**Comment by Mingrg** [comment]
Hi, @kan-bayashi 
    I tried to train a fastspeech2 without MFA, it's not bad, and not so good. Because it has some noise, I'm trying to figure it out why that noise occurred? And I will update my experiment result in the future.

Now, I have another question:
`In fastspeech2 training. Do I have to use normalization?`
Such as in conf/tuning/train_fastspeech2.yaml.
- encoder_normalize_before=true
- decoder_normalize_before=ture
- pitch_normalize=global_mvn
- energy_normalize=global_mvn

    Is this 4 normalizations must do? If I don't use this 4 normalizations, whether the performance will be much more bad? I'm not sure have you did that before :)

**Comment by kan-bayashi** [comment]
As you can see the comment in the config, 
- `encoder_normalize_before=true`
- `decoder_normalize_before=ture`

mean whether to perform layer normalization at first in the transformer block.
This helps to robustly learn the diagonal attention in AR models.
For non-AR models, I think this parameter make the training more robust to the optimizer setting (e.g. lr, batchsize, etc.)

`pitch_normalize=global_mvn` and `energy_normalize=global_mvn` make the scale of feature to be mean 0 variance 1.
I've never tried without this normalization in FastSpeech2, but it might balance the scale of each loss function (pitch loss, energy loss, mel loss) since pitch and energy have different value range.
